services/horizon/internal/ingest/processors/operations_processor.go

```diff
@@ -655,6 +705,64 @@ func (operation *transactionOperationWrapper) Details() (map[string]interface{},
+// SAC events that involve asset balance changes follow some standard data formats:
+//     amount expressed as uint64 only, the event type in this case provides the context of whether an amount was credit/debit to a balance.
+//     the 'from' and 'to' attributes represent an account or a contract
+func (operation *transactionOperationWrapper) parseAssetBalanceChangesFromContractEvents() ([]map[string]interface{}, error) {
+	balanceChanges := []map[string]interface{}{}
+
+	txMeta, ok := operation.transaction.UnsafeMeta.GetV3()
+	if !ok {
+		// there's no newer v3 meta which is when contract events start being present, otherwise none present.
+		return balanceChanges, nil
+	}
+
+	for _, opEvents := range txMeta.Events {
+		for _, contractEvent := range opEvents.Events {
+			// parse the xdr contract event to contractevents.StellarAssetContractEvent model
+			// has some convenience like to/from attributes are expressed in strkey format for contracts(C...) and accounts(G...)
+			if sacEvent, err := contractevents.NewStellarAssetContractEvent(&contractEvent, operation.networkPassphrase); err == nil {
+				switch sacEvent.GetType() {
+				case contractevents.EventTypeTransfer:
+					transferEvt := sacEvent.(contractevents.TransferEvent)
+					balanceChanges = append(balanceChanges, createSACBalanceChangeEntry(transferEvt.From, transferEvt.To, transferEvt.Amount.Lo, transferEvt.Asset))
+				case contractevents.EventTypeMint:
+					mintEvt := sacEvent.(contractevents.MintEvent)
+					balanceChanges = append(balanceChanges, createSACBalanceChangeEntry(mintEvt.Admin, mintEvt.To, mintEvt.Amount.Lo, mintEvt.Asset))
+				case contractevents.EventTypeClawback:
+					clawbackEvt := sacEvent.(contractevents.ClawbackEvent)
+					balanceChanges = append(balanceChanges, createSACBalanceChangeEntry(clawbackEvt.From, clawbackEvt.Admin, clawbackEvt.Amount.Lo, clawbackEvt.Asset))
+				case contractevents.EventTypeBurn:
+					burnEvt := sacEvent.(contractevents.BurnEvent)
+					balanceChanges = append(balanceChanges, createSACBalanceChangeEntry(burnEvt.From, "", burnEvt.Amount.Lo, burnEvt.Asset))
+				}
+			}
+		}
+	}
+
+	return balanceChanges, nil
+}
+
+// fromAccount - strkey format of contract or address
+// toAccount - strkey format of contract or address, or nillable
+// amountChanged - absolute value that asset balance changed
+// asset - the fully qualified issuer:code for asset that had balance change
+//
+// return - a balance changed record expressed as map of key/value's
+func createSACBalanceChangeEntry(fromAccount string, toAccount string, amountChanged xdr.Uint64, asset xdr.Asset) map[string]interface{} {
+	balanceChange := map[string]interface{}{}
+
+	balanceChange["from"] = fromAccount
+	if toAccount != "" {
+		balanceChange["to"] = toAccount
+	}
+	balanceChange["amount"] = string(amountChanged)
```

Comment:
bit of a nit: any reason not to initialize it directly with the "known" fields? e.g.
```suggestion
	balanceChange := map[string]interface{}{
		"amount": string(amountChanged),
		"from": fromAccount,
	}
	if toAccount != "" {
		balanceChange["to"] = toAccount
	}
```

Reply:
I find it easier to visualize the structure this way rather than adding entries one by one in my head

Reply:
sure, may be a little more fluent.